[WIP][AMDGPU] Improve the handling of `inreg` arguments

shiltian opened the pull request:
When SGPRs available for `inreg` argument passing run out, the compiler silently
falls back to using whole VGPRs to pass those arguments. Ideally, instead of
using whole VGPRs, we should pack `inreg` arguments into individual lanes of
VGPRs.

This PR introduces `InregVGPRSpiller`, which handles this packing. It uses
`v_writelane` at the call site to place `inreg` arguments into specific VGPR
lanes, and then extracts them in the callee using `v_readlane`.

Fixes #130443 and #129071.

shiltian commented:
* **#133614** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133614?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133614?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shiltian commented:
I'm still working on this, but I'd like to get some early feedback. The newly added test case currently crashes in the verifier with the following error:

```
error: Illegal instruction detected: Illegal immediate value for operand.
renamable $vgpr20 = V_WRITELANE_B32 killed $sgpr0, killed $sgpr1, 0
error: Illegal instruction detected: Illegal immediate value for operand.
renamable $vgpr1 = V_WRITELANE_B32 killed $sgpr0, killed $sgpr1, 1
```

Looks like I can't just use an immediate operand for the lane selection. I'll fix that in a later update.

shiltian commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,86 @@ void SITargetLowering::insertCopiesSplitCSR(
+  MachineFunction &MF;
+
+  Register getCurReg() const { return CurReg; }
+  unsigned getCurLane() const { return CurLane; }
```
This should be `CurLane % WaveFrontSize`.

shiltian commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,86 @@ void SITargetLowering::insertCopiesSplitCSR(
   }
 }
 
+class InregVPGRSpiller {
```
Need document here.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,86 @@ void SITargetLowering::insertCopiesSplitCSR(
+
+    MF.addLiveIn(getCurReg(), &AMDGPU::VGPR_32RegClass);
+
+    // TODO: Do we need the chain here?
```
Yes. You should be using getCopyFromReg instead of directly referencing getRegister

shiltian replied:
But IIUC we can't simply call `getCopyFromReg` right? We want to copy from the register of exact lane, and that's why we need to use the readlane intrinsic here.

arsenm replied:
It's 2 steps. getCopyFromReg to get the physical snapshot at the ABI boundary, and then the readlane on the value to extract the lane from the virtual register copy

shiltian commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,86 @@ void SITargetLowering::insertCopiesSplitCSR(
+
+    SmallVector<SDValue, 4> Operands{
+        DAG.getTargetConstant(Intrinsic::amdgcn_writelane, SL, MVT::i32),
+        DAG.getRegister(getCurReg(), VT), V,
```
@arsenm What do we need to do here? Do we need to call `getCopyToReg` here? If we do, doesn't it end up with writing different values to the same register?

arsenm replied:
Yes, it's the same as the input case just the opposite. Build the vreg value and one final copyToReg for the physical register

shiltian commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,91 @@ void SITargetLowering::insertCopiesSplitCSR(
+        DAG.getTargetConstant(Intrinsic::amdgcn_writelane, SL, MVT::i32),
+        DAG.getRegister(DstReg, VT), Val,
+        DAG.getTargetConstant(CurLane++, SL, MVT::i32)};
+    return {DAG.getNode(ISD::INTRINSIC_WO_CHAIN, SL, VT, Operands), Glue};
```
@arsenm I'm not sure how to handle this part correctly. Basically we want to return something similar to `DAG.getCopyToReg`, because in the normal case it has the following handling:

```
      Chain = DAG.getCopyToReg(Chain, DL, Reg, Val, InGlue);
      InGlue = Chain.getValue(1);
```

I tried something like the following:

```
  SDValue write(SDValue Chain, const SDLoc &SL,
                                    Register &Reg, SDValue Val, SDValue InGlue,
                                    EVT VT) {
    if (DstReg.isValid()) {
      Reg = DstReg;
    } else {
      DstReg = Reg;
      DstVal = DAG.getCopyToReg(Chain, SL, Reg, Val, InGlue);
    }
    // According to the calling convention, only SGPR4–SGPR29 should be used for
    // passing 'inreg' function arguments. Therefore, the number of 'inreg' VGPR
    // arguments must not exceed 26.
    assert(CurLane < 26 && "more than expected VGPR inreg arguments");
    SmallVector<SDValue, 4> Operands{
        DAG.getTargetConstant(Intrinsic::amdgcn_writelane, SL, MVT::i32),
        DAG.getRegister(DstReg, VT), Val,
        DAG.getTargetConstant(CurLane++, SL, MVT::i32)};
    DAG.getNode(ISD::INTRINSIC_WO_CHAIN, SL, VT, Operands);
    return DstVal;
  }
```

And then:

```
    if (ArgIdx >= NumSpecialInputs &&
        Outs[ArgIdx - NumSpecialInputs].Flags.isInReg() &&
        AMDGPU::VGPR_32RegClass.contains(Reg)) {
      Chain = Spiller.write(Chain, DL, Reg, Val, InGlue,
                            ArgLocs[ArgIdx - NumSpecialInputs].getLocVT());
    } else {
      Chain = DAG.getCopyToReg(Chain, DL, Reg, Val, InGlue);
    }
    InGlue = Chain.getValue(1);
```

This doesn't seem to work because there doesn't seem to be an actual "effect" of the write lane intrinsic and then it is simply not emitted, as shown in the newly added test.

The code currently in the PR doesn't work as well, because the same glue on the different values messes up the scheduling.

shiltian commented on llvm/test/CodeGen/AMDGPU/inreg-vgpr-spill.ll:
```diff
@@ -0,0 +1,107 @@
+; CHECK-NEXT:    s_addc_u32 s9, s5, 0
+; CHECK-NEXT:    v_mov_b32_e32 v1, v0
+; CHECK-NEXT:    s_waitcnt lgkmcnt(0)
+; CHECK-NEXT:    v_writelane_b32 v1, s30, 0
```
There is some sort of mismatch here. It writes to `v1` at call site, but reads `v0` in the callee.

shiltian replied:
@arsenm The issue is here. We use `v1` for the spilled `inreg` argument passing, but in the callee, it is `v0`. However, the physical register at the call site we are using is `v0`.

shiltian replied:
@arsenm ping, this is for https://github.com/llvm/llvm-project/pull/133614#discussion_r2041136303

shiltian commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,96 @@ void SITargetLowering::insertCopiesSplitCSR(
+        DAG.getTargetConstant(Intrinsic::amdgcn_writelane, SL, MVT::i32), Val,
+        DAG.getTargetConstant(CurLane++, SL, MVT::i32)};
+    if (!LastWrite) {
+      Register VReg = MF.getRegInfo().getLiveInVirtReg(DstReg);
```
The mismatch is from here: the first tied-input attached to the intrinsic. Basically we want the intrinsic to write to a specific register. The input `Reg` is actually a `MCRegister`. We need a virtual register here, therefore we use `getLiveInVirtReg` first to get a virtual register corresponds to the `MCRegister`, and then use `getRegister` to get an `SDValue`. Thhe generated MIR would be something like below:

```
%0:vgpr_32(s32) = COPY $vgpr0
...
%33:vgpr_32 = V_WRITELANE_B32 killed %28:sreg_32, 0, %0:vgpr_32(tied-def 0)(s32)
%34:vgpr_32 = V_WRITELANE_B32 killed %29:sreg_32, 1, %33:vgpr_32(tied-def 0)
```

This sequence later is lowered to something like below:

```
%34:vgpr_32 = COPY %0:vgpr_32(s32)
%34:vgpr_32 = V_WRITELANE_B32 $sgpr30, 0, %34:vgpr_32(tied-def 0)
%34:vgpr_32 = V_WRITELANE_B32 $sgpr31, 1, %34:vgpr_32(tied-def 0)
```

And further lowered to the following after RA:

```
renamable $vgpr1 = COPY renamable $vgpr0
renamable $vgpr1 = V_WRITELANE_B32 $sgpr30, 0, killed $vgpr1(tied-def 0)
renamable $vgpr1 = V_WRITELANE_B32 $sgpr31, 1, killed $vgpr1(tied-def 0)
```

As we can see here, we are supposed to write to `$vgpr0`, but because the `COPY` reads from `$vgpr0` to `$vgpr1`, and then `V_WRITELANE_B32` becomes write to `$vgpr1`.

We probably should not use virtual register as the operand of the intrinsic, but that will crash the two address instruction pass, which expects the operand to be a vreg.

arsenm replied:
> We probably shouldn't be using virtual registers as operands directly in the intrinsic,

The intrinsics should *only* use virtual registers.

> %34:vgpr_32 = COPY %0:vgpr_32(s32)

The problem here is we probably should be using a WWM copy, not a regular one

arsenm replied:
This example MIR is missing the output part. This is all fine as long as it copies to the correct register in the end. I don't see what the issue is?

arsenm commented:
Vaguely related, but we should also support inreg on function returns so we can avoid using the ugly hack of integer typed return values with shader calling conventions. Still not as nice as a proper per-field entry on the struct

shiltian commented:
> Vaguely related, but we should also support inreg on function returns so we can avoid using the ugly hack of integer typed return values with shader calling conventions. Still not as nice as a proper per-field entry on the struct

That'd be in the follow up PRs.

shiltian commented:
gentle ping

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,96 @@ void SITargetLowering::insertCopiesSplitCSR(
+      : State(State), DAG(DAG), MF(MF) {}
+
+  SDValue readLane(SDValue Chain, const SDLoc &SL, Register &Reg, EVT VT) {
+    if (SrcVal) {
+      State.DeallocateReg(Reg);
+    } else {
```
We shouldn't really be deallocating anything. The CC assignment should have directly returned return in vN at index M. Should be able to do this by using custom allocation type

shiltian replied:
> The CC assignment should have directly returned return in vN at index M.

How to express it though? Like how to refer to VGPR N at lane M? It doesn't seem like we can do it at the moment but maybe to add it in the td file, like we have the all possible register sequences?

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2841,6 +2841,96 @@ void SITargetLowering::insertCopiesSplitCSR(
+  SDValue finalize(SDValue Chain, const SDLoc &SL, SDValue InGlue) {
+    if (!LastWrite)
+      return LastWrite;
+    return DAG.getCopyToReg(Chain, SL, DstReg, LastWrite, InGlue);
```
Unfortunately I think we need to use a WWM copy here. Unfortunately that would also break any of the coalescing we would want in the common case unless we do something about that

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -3130,8 +3221,17 @@ SDValue SITargetLowering::LowerFormalArguments(
+      Val = Spiller.readLane(Chain, DL, Reg, VT);
+    } else {
+      Reg = MF.addLiveIn(Reg, RC);
+      Val = DAG.getCopyFromReg(Chain, DL, Reg, VT);
```
will probably also need to be a WWM copy

shiltian committed:
[AMDGPU] Pack overflow inreg args into VGPR lanes

When inreg function arguments overflow the available SGPRs, pack multiple values
into lanes of a single VGPR using writelane/readlane instead of consuming one
VGPR per overflow argument.

The feature is behind a flag (default off) and currently only supports the
SelectionDAG path.

Known issue: if the register allocator does not coalesce the COPY between the
writelane chain and the physical call argument register, the resulting v_mov_b32
is EXEC-dependent and will not transfer inactive lanes. This is correct when
EXEC is all-ones (the common case at call sites) but would be incorrect inside
divergent control flow.

shiltian commented on llvm/test/CodeGen/AMDGPU/inreg-vgpr-lane-packing.ll:
```diff
@@ -0,0 +1,282 @@
+; CHECK-NEXT:    s_mov_b32 s28, 12
+; CHECK-NEXT:    s_mov_b32 s29, 13
+; CHECK-NEXT:    v_mov_b32_e32 v1, v0
+; CHECK-NEXT:    v_mov_b32_e32 v0, v2
```
Here it shows the issue when coalescing fails, we might run into issue, because at call site, the writelane writes to v2 followed by a copy from v2 to v0, and then in the callee, the readlane directly reads from v0. If the caller is in a divergent branch, then this would not correctly.

However, the challenging problem is, how can we issue a WWM_COPY at ISel stage? Maybe we need to add a pseudo node and then change it to something else later? The current WWM_COPY rewriting pass seems like only tailored for VGPR spilling.

shiltian reviewed (commented):
This PR now contains changes from all the three PRs because Graphite doesn't allow to retarget a PR.

shiltian commented:
It doesn't seem like this optimization can be used at a `tail call` call site, thus we will need to check if a function is only called in a regular call site. This is not something that can be done at function lowering stage.

arsenm commented:
> It doesn't seem like this optimization can be used at a `tail call` call site, thus we will need to check if a function is only called in a regular call site. This is not something that can be done at function lowering stage.

I don't see why not? The argument list is the same either way

jayfoad commented:
> When SGPRs available for `inreg` argument passing run out, the compiler silently
> falls back to using whole VGPRs to pass those arguments. Ideally, instead of
> using whole VGPRs, we should pack `inreg` arguments into individual lanes of
> VGPRs.

This sounds like a bad idea. How do you know it's safe to clobber lane 0 of some VGPR, when lane 0 might not be active? Would you introduce a whole wave spill of the VGPR?

arsenm commented:
> This sounds like a bad idea. How do you know it's safe to clobber lane 0 of some VGPR, when lane 0 might not be active? Would you introduce a whole wave spill of the VGPR?

Yes, this needs to use WWM values

jayfoad commented:
> > This sounds like a bad idea. How do you know it's safe to clobber lane 0 of some VGPR, when lane 0 might not be active? Would you introduce a whole wave spill of the VGPR?
> 
> Yes, this needs to use WWM values

OK then it is not at all clear that this is a desirable change to the ABI. It seems like it could make the code perform worse in a lot of cases.
